Fixlenspolarizationrotation

briandleahy opened the pull request:
Fixes 371.

This fixes two bugs with the implementation of polarization rotation in the `Lens` theory. These issues were only present when the polarization was not along the x (or [1, 0]) direction. With this PR, the agreement between MieLens and Lens(Mie) does not change with polarization, and both theories produce rotationally-invariant results for rotationally-invariant objects (i.e. the parallel component of the field scattered from a sphere, evaluated at a position that is along the polarizazion vector, does not change as the polarization changes; this was always true for MieLens but was not true for Lens(Mie)).

Note that this does not change the calculations at all when the polarization is along the x-direction, so #350 is still open.

briandleahy commented:
@ralex0 -- I made some slight edtis to the `numexpr` usage, but I don't have `numexpr` working on my machine. Can you check that the `numexpr` and no-numexpr pathwasy produce the same result?
@barkls -- as we discussed on Thurs, I also threw in 2 minor fixes with this PR (faster test_interface and making the DDA tests pass). However I'm not sure if the DDA tolerances should be changed. If you want, we can leave the DDA tests as failing for this PR. Let me know.

barkls commented:
Looks good, I just have a few additional comments:

1) It would be nice to have a short summary in this thread of what the bugs were for future reference, especially since the minor rewrite of the Lens class makes it hard to pinpoint from the code.

2) Please update the changelog.

briandleahy commented on holopy/scattering/tests/test_lens.py:
```diff
@@ -64,10 +67,10 @@ def test_integrate_over_phi_with_quad_points(self):
 
         def func(x):
-            return np.cos(np.pi * x)
+            return np.exp(-3 * np.sin(x))
```
The old quadrature was optimized for arbitrary functions on an interval, using Gauss-Legendre points and weights. However, these are not optimal for periodic functions, and the actual integrals we are evaluating are periodic with period 2 pi. So I optimized the phi quadrature for a function with period 2 pi, which gives a 3x improvement in accuracy with the same computational cost.

However, the old function in the test, cos (pi * x), is not 2 pi - periodic, and therefore the quadrature does not perform as well on this function as it did before, and the test did not pass. This is a problem with the test and not the code, so I changed the test. I could have changed it to just cos(x), but  that is a more well-behaved function. Since I needed to change the function, I figured that I might as well make the test more challenging as well.

barkls replied:
thanks for clarifying!

briandleahy commented on holopy/scattering/theory/dda.py:
```diff
@@ -82,7 +82,7 @@ def __init__(self, n_cpu = 1, max_dpl_size=None, use_indicators=True,
         # Check that adda is present and able to run
         try:
-            with SuppressOutput():
+            with SuppressOutput(suppress_output=suppress_C_output):
```
For some reason the `SuppressOutput`still causes crashes in ipython terminals on my machine, and this was stopping me from running DDA in an ipython shell.  At some point I should come back to the `SuppressOutput`ipython crash.

briandleahy commented:
> 1. It would be nice to have a short summary in this thread of what the bugs were for future reference, especially since the minor rewrite of the Lens class makes it hard to pinpoint from the code.

There were two separate problems, both just involving polarization rotations. 

The first problem was that the code rotated the object (by rotating the calculation of the scattering matrices) when the polarization was rotated. This was a simple bugfix.

The second problem was a more subtle error, corresponding to a math error in rotating the polarization (so apologies for having to read it in markdown). Take a look at Eq. 1 from the perspective paper below, which is correct for x-polarized light:
![Screenshot from 2020-10-26 11-38-36](https://user-images.githubusercontent.com/25911342/97193850-d95d7c00-177f-11eb-87d3-2faa7cc2a717.png)
Previously, the `Lens` theory was calculating the scattered fields by assuming that the component of the scattered field parallel to the polarization was the same, independent of whether the light was x-polarized or polarized in an arbitrary direction. This is equivalent to changing every "x-hat" in Eq. (1) into a "p-hat", where "p-hat" is the polarization direction (and "y-hat" with a vector perpendicular to the polarization). This is also incorrect, because the "x-hat"s appear for two separate reasons. The terms **theta** * **S** * **x** and **phi** * **S** * **x** capture the effects of the polarization of the incoming beam, and need to rotate when the polarization changes. But the terms (**x** cos(phi) + **y** sin(phi)) etc describe theta and phi unit vectors on the lens pupil, and do not rotate when the polarization changes. This is a conceptual difference, but it ends up being a small code change at the end -- the integrands just get a cos(phi - pol_angle) and sin(phi - pol_angle) instead of cos(phi) and sin(phi), lines 141-142 and 151-152. 

If you want I can quickly TeX up the derivation and place it somewhere, perhaps here or on the group wiki.

pep8speaks commented:
Hello @briandleahy! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

















* In the file [`holopy/scattering/theory/lens.py`](https://github.com/manoharan-lab/holopy/blob/4c63307ba39365d5fcfdde310b14b9d38ec62c2f/holopy/scattering/theory/lens.py):

> [Line 163:30](https://github.com/manoharan-lab/holopy/blob/4c63307ba39365d5fcfdde310b14b9d38ec62c2f/holopy/scattering/theory/lens.py#L163): [E201](https://duckduckgo.com/?q=pep8%20E201) whitespace after '['

briandleahy commented on holopy/scattering/tests/test_lens.py:
```diff
@@ -248,6 +214,136 @@ def test_numexpr_integrandr(self):
 
+    @attr('medium')
+    def test_polarization_rotation_produces_small_changes_to_image(self):
+        # we test that, for two sphere, rotating the polarization is
```
Fixed by 4c63307b

briandleahy commented on holopy/scattering/tests/test_lens.py:
```diff
@@ -248,6 +214,136 @@ def test_numexpr_integrandr(self):
+        intensity_xpol = np.linalg.norm(fields_xpol, axis=0)**2
+        intensity_ypol = np.linalg.norm(fields_ypol, axis=0)**2
+
+        # We are just trying to cehck that rotating the polarization
```
Fixed by 4c63307b

briandleahy commented:
> 2. Please update the changelog.

I've also added a line to the changelog, in 4c63307b

briandleahy commented:
@barkls -- let me know if I missed anything or if this is ready to merge.

barkls commented:
I think we're just waiting on a test of numexpr from @ralex0

briandleahy commented:
@barkls -- I installed `numexpr` and checked this on my local machine. There were some issues, which I fixed. Those changes are now pushed (primarily in 7e63fc92 ). Please take a look and let me know if something seems off!

I did check that changing the `numexpr` expressions causes the tests to fail. I also adjusted the tests to compare the numexpr pathway against the numpy pathway, which in my opinion is what matters anyways.

barkls commented on holopy/scattering/tests/test_lens.py:
```diff
@@ -155,64 +150,63 @@ def test_calc_holo_theta_npts_not_equal_phi_npts(self):
+
+        prefactor_numexpr = LENSMIE._integrand_prefactor(krho, phi, kz)
+
+        # Monkey-patch lens to use the numpy pathway:
```
I think this is a sign that some functionality is missing - e.g. calling Lens with a `computation_method='auto'` that can be overridden to specify numpy or numexpr. 

However, we also want to avoid complexifying [not a word] the interface, and there's no real reason you'd want to enforce numpy outside of testing so this seems fine.

barkls replied:
I notice you do this a few times through this test file. You could consider writing a context manager for the monkey patch. That would also make it less likely someone breaks things later since the numexpr adjustments would be removed from the tests themselves.

briandleahy replied:
Should be fixed by babd9bcd. Let me re-run the tests quickly on my computer with numexpr working before merging this in though.

barkls commented:
Thanks for following up on this @briandleahy and patching things to work with numexpr. I suggested a context manager to handle the numexpr monkey-patches, but we can merge in as is if you don't feel like doing that.

At some point it would be nice to get things like DDA and numexpr tested on Travis, but that's definitely beyond the scope of this PR.

briandleahy commented:
Tests work with new commit. I'm ready to merge this in when you are.

barkls commented:
good fix!
